EMAILPLT-190 Refactor styles and update jQuery, jQuery UI and fluid

cousquer opened the pull request:
**Refactoring of Email-preview Portlet and Library updates**
more informations and screencap at : 
[https://issues.jasig.org/browse/EMAILPLT-190](https://issues.jasig.org/browse/EMAILPLT-190)

3 breackpoints : mobile, tablet, desktop
- Double inner offcanvas Options&folder block (offcanvas1)- Emailmessages&toolbar&pagination bock - selected email block (offcanvas2)
- fix select all-checkboxes checkbox
- transform allFolder select into a list of button (with unread message badge) with Quick[select](MIT Licence) [https://github.com/eggboxio/quick-select](https://github.com/eggboxio/quick-select) - [http://quick-select.eggbox.io/](http://quick-select.eggbox.io/)
- improvement in i18n and l10n
- Accessibility improvements
- keyboard navigation with tab, space and enter keys on windows and linux
- update to fluid 1.4 -> done
- update of jquery, jquery UI
- use the same less mixin than uPortal less mixin -> it will be easy to update bootstrap to 4 and SASS or LESS, or change to Material
- Fluid Accessibility issue FLUID-6064 with role="application" Solved and patched 😄 -> overrided to role="section" at hot.
- Add comments following Google JavaScript Style Guide & LESS following code styleguide conventions

The update to fluid 1.5 is in progress (WIP). it will be in another PR after.
The implementation of mobile view in desktop multicolumns dashboard will be in another PR.
There 's a lot of Accessibility Improvements : keyboard navigation, aria attributes additions, etc.

test ok with uPortal.4.3.+ with gmail and a linux webmail.

tested on : 
working on old version of IOS (IOS 5.5) (tests ok with iPad v1...)
IOS : safari FF, Chrome without/with VoiceOver 
Android : chrome, FF without/with talkback
Win10 : chrome, FF, safari, edge, ie11, ie10
OSX : chrome, safari, FF  without/with VoiceOver

cousquer commented:
Many thanks to Marissa Warner-Wu of the University of Edinburgh for her advices on the User eXperience at Apereo2016 NY.

cousquer commented:
I've remove some quibbles around fluid version (now it is the regular fluid.1.4.0 - it seems that [https://issues.jasig.org/browse/NEWSPLT-65](url) is resolved. I can't reproduce now the double escaping by fluid and antisamy).
Jquery-migrate is needed because fluid 1.4.0 need "browser.msie". Without migrate you've got a uncaught TypeError.

bjagg commented on src/main/webapp/WEB-INF/jsp/editPreferences.jsp:
```diff
@@ -422,4 +425,4 @@
     });
 
-</script>
+</rs:compressJs></script>
```
Thank you for leveraging compression here!

bjagg commented on src/main/webapp/js/email-browser.js:
```diff
@@ -31,17 +99,18 @@ var jasig = jasig || {};
         var messageId = $(element).attr("messageId");
         var message = getMessage(that, messageId);
 
+      if(message) {
```
Space between "if" and "("?

bjagg commented on src/main/webapp/js/jquery.quickselect.js:
```diff
@@ -0,0 +1,143 @@
+			});
+
+
+			if(breakOutAll) {
```
Missing space after "if"?

bjagg commented:
Excellent, substantial work, Christian!  I have nitpicked a few things. Appreciate you setting the bar high.

cousquer commented:
@bjagg,
I've fixed all the missing spaces after if in js, added a comment in messages.properties and inhibited specific jQmobile view of mUniversality

ChristianMurphy commented on src/main/webapp/js/email-browser.js:
```diff
@@ -19,6 +19,74 @@
 
 (function($, fluid) {
+
+	var ischecked = 0;
```
is `isChecked` counting the number of checks?
To me `is<something>` expressions are usually boolean.

cousquer replied:
@ChristianMurphy,

> is  `isChecked` counting the number of checks?

Yes, because I need to catch the global checkness state of every checkboxes in the Messages List. To handle every cases in Tablet's / Desktop's View :

- when  a user selects one or many checkboxes -> `.toolbar-middle` slides down and the user can suppress or not the selected message(s) or can selected all the displayed messages by selecting checkbox of `.toolbar-middle` ;

- when  a user has selected one or many checkboxes and he clicks on a message to display it -> because there is a suppress button on the displayed message, I flush `ischecked` to 0 to slide up `.toolbar-middle...`

So I've got the same wondering as you when I've created 'isChecked', but because I'm not a native english speaker (:tada: :fr:  :wine_glass: :confetti_ball:) , at this time :sweat_smile:, I did not know how to name simply `the counting of the checkboxesGlobalChecknessState`... If you've got a idea of something meanfull for us, please help me. :simple_smile:

ChristianMurphy replied:
> I did not know how to name simply `the counting of the checkboxesGlobalChecknessState`

A little long, but you pretty much got it. :+1:  :smile: :gb: 
`checkboxCounter` could be shorter alternative.

cousquer replied:
👍 for checkboxCounter + a comment on checkboxes Global Checkness State with [google style guidlines](https://google.github.io/styleguide/javascriptguide.xml?showone=Comments#Comments)

ChristianMurphy commented on src/main/webapp/js/email-browser.js:
```diff
@@ -252,15 +373,35 @@ var jasig = jasig || {};
 
         var batchOptions = {
-            batchSize: that.options.batchSize,
+            batchSize: that.options.batchSize,            
```
:art: remove trailing spaces

ChristianMurphy commented on src/main/webapp/js/email-browser.js:
```diff
@@ -357,9 +528,22 @@ var jasig = jasig || {};
+        // Notify the server of changes to pageSize so they can be remembered and update the container min-height class
+        that.pager.pager.events.initiatePageSizeChange.addListener(function(newPageSize) {
+            // console.log("newPageSize: "+ newPageSize);
+            /* if #right-content-email has .emailpreview-container-min-height-XX class remove it */
```
would using `.emailpreview-container.min-height-xx` rather than `.emailpreview-container-min-height-xx` simplify this logic any? or would it be about the same?

cousquer replied:
:+1:

cousquer replied:
@ChristianMurphy 
it will be #right-content-email.min-height-XX
according to my comment in [theme.less#L343](https://github.com/cousquer/email-preview/blob/3a68f3413c79609733d7c2ab7d9c2977980302f3/src/main/webapp/less/includes/theme.less#L343)

ChristianMurphy commented on src/main/webapp/less/includes/theme.less:
```diff
@@ -16,79 +16,1048 @@
+                color: @white !important;
+                cursor: default;
+                z-index: 2;
+            }
```
The lines 300-464 look to have well qualified selectors, what's causing the conflict that `!important` is needed?

cousquer replied:
@ChristianMurphy 
300-464 : converts Selected Message's "fixed content" to "fluid content"
I know that `!important` is a question, but when you've got newsletters designers who are using `nested tables with html inlines styles***` such : 

`<TABLE WIDTH="1024" HEIGHT="960"><TR><TD><TABLE WIDTH="500" HEIGHT="30">...bunch of nested tables`

`***` it's the only ways of doing things in html email rendering...

You end up with something like this :

![alt text][img]

[img]: https://issues.jasig.org/secure/attachment/15416/emailpreview-displaying-none-fluid-message-eg-a%20newsletter-beforetheCSSimportantpatch.png "The fixed newsletter overflow crappy display"

but with !important statement (it's scoped only on the message div), I've got this:


![alt text][img2]

[img2]: https://issues.jasig.org/secure/attachment/15246/fixedcontentConverttoFluid%20content.png "The fixed newsletter convert to fluid display"

without injecting any parsing JS that need to parse the message content which will be **a security and integrity issue** :warning: :collision: :no_entry_sign: ...
for the moment, I know that Google's statement for removing `!important` of any CSS code is an advice and because there are someothers none [Blink ](https://en.wikipedia.org/wiki/Blink_(web_engine))rendering engine like Chrome that do not have the same selectors priority hierarchy, like [Webkit ](https://en.wikipedia.org/wiki/WebKit)Safari I do not mention [Trident](https://en.wikipedia.org/wiki/Trident_(layout_engine)), `!important` is still very usefull in some cases... :smiley:
 Trust me... :wink:

ChristianMurphy replied:
Sounds good just wanted to double check. :+1:

cousquer replied:
@ChristianMurphy 
I will push tomorrow the fixes. and give a try to override role="application" by role="section" and see how and what [that](https://github.com/fluid-project/infusion/blob/a69345054ccd90787274b55e7943940ffe4572ef/src/components/pager/js/Pager.js#L509) will react...

ChristianMurphy commented on src/main/webapp/js/email-browser.js:
```diff
@@ -19,6 +19,74 @@
+                
+	function hideAndShowMessage() {
+        	var linkMessages = $('#email-list-table .right-content-email-toggle .subject');
+        	for (var i=0; i < linkMessages.length; i++) {
```
🎨 `for (var i = 0; ...`

cousquer replied:
ok

ChristianMurphy commented on src/main/webapp/js/email-browser.js:
```diff
@@ -489,39 +713,56 @@ var jasig = jasig || {};
+	            	getEmail(that, 0, that.options.batchSize, undefined, undefined, param1);
 	            	location.reload();
-            	});             
+            	});    
```
🎨 remove trailing spaces

ChristianMurphy commented on src/main/webapp/js/email-browser.js:
```diff
@@ -420,61 +604,101 @@ var jasig = jasig || {};
+                var chk2 = $(this).prop('checked');
+                // console.log("chk= "+ chk);
+                // console.log("chk2= "+ chk2);
+                if (chk2 == true) {
```
🎨 `if (chk2) {`

ChristianMurphy commented on src/main/webapp/js/email-browser.js:
```diff
@@ -420,61 +604,101 @@ var jasig = jasig || {};
 		                var selected;
 		                $.each(response, function(index,response) {
-		                	if(index=="selected"){
+		                	if (index=="selected"){
```
🎨 `if (index === 'selected') {`

cousquer replied:
:+1: 
:flushed:, How could I have let this in the code... 

> if ( :beer: || :sake: || :cocktail: || :tropical_drink: || :wine_glass: ) {
   goToSleep();
 } else {
   goCoding();
}

ChristianMurphy commented on src/main/webapp/js/email-browser.js:
```diff
@@ -489,39 +713,56 @@ var jasig = jasig || {};
 
         return that;
-
+        
```
🎨 remove trailing spaces

ChristianMurphy commented:
> Some issues due to `role="application"` added by fluid complicate the navigation for blind people on IOS iphone and Mac

This will probably need to be resolved upstream first.
Here is offending line of code.

https://github.com/fluid-project/infusion/blob/a69345054ccd90787274b55e7943940ffe4572ef/src/components/pager/js/Pager.js#L505

cousquer commented:
@ChristianMurphy 
have a look at this override : 
[https://github.com/Jasig/email-preview/pull/70/files#diff-77d9fb9013fb79d20944f68094933e90R378](https://github.com/Jasig/email-preview/pull/70/files#diff-77d9fb9013fb79d20944f68094933e90R378)

and at this (a few lines on top) : 
[https://github.com/fluid-project/infusion/blob/a69345054ccd90787274b55e7943940ffe4572ef/src/components/pager/js/Pager.js#L459](https://github.com/fluid-project/infusion/blob/a69345054ccd90787274b55e7943940ffe4572ef/src/components/pager/js/Pager.js#L459)

maybe it can be patch first on our side.

cousquer commented:
@ChristianMurphy 
I've Pushed the fixes of your review.
dig now into this role="application" question...
Thank's so much.

cousquer commented:
Do not merge, I will add tonight CSS3 z-index and shadow effects ala MaterialDesign to have a visual track of the Fix to Fluid issue. :wink:

cousquer commented:
@ChristianMurphy, @bjagg , @drewwills 
[https://issues.fluidproject.org/browse/FLUID-6064](https://issues.fluidproject.org/browse/FLUID-6064) is now patched in this PR. I've put references url in the Js comments.
Add also shadow and z-index effects ala MaterialDesign to have a visual track.

ChristianMurphy commented on src/main/webapp/js/email-browser.js:
```diff
@@ -55,6 +55,23 @@ var jasig = jasig || {};
+         *    Additional notes: http://tink.uk/using-the-aria-application-role/
+         * @supported So far tested in Chrome, FF, Safari, IE9, edge. IOS, Android.
+         */
+        function overrideA11nRole (elem, actualRole, newRole) {
```
This patch looks okay for now.
Later this week I'll look into getting a PR into fluid infusion itself, so the fix could be a version change of fluid instead of DOM replacement in uPortal.

cousquer replied:
push on 1.4.0 also

ChristianMurphy replied:
Opened PR to patching FLUID 1.4 https://github.com/fluid-project/infusion/pull/785

ChristianMurphy commented on src/main/webapp/less/includes/theme.less:
```diff
@@ -477,6 +482,14 @@
             padding: 7px 10px !important;
+            &:hover {
+                box-shadow: 5px 6px 7px 0 rgba(0, 0, 0, 0.2), 0 1px 2px 0 rgba(0, 0, 0, 0.1);
+                -moz-transition:all .3s ease-in-out;
```
:art: `.3` -> `0.3`

cousquer replied:
ok

cousquer commented:
@ChristianMurphy 
I've got a funny [3.1.4 Abbreviations Level AAA](https://www.w3.org/WAI/WCAG20/quickref/#meaning-located) issue with voiceOver due to gmail with email-quota-limit : :laughing:
google sends me 15.00GB and voiceOver spells 

> 1 - 5 - dot - 0 - 0 - :gb:

ChristianMurphy commented:
The `role="application"` has been removed from pager on master branch of fluid infusion.
A related issue on pager was found while testing the `role="application"` changes that *may* affect email preview.
https://issues.fluidproject.org/browse/FLUID-6081

cousquer commented:
@ChristianMurphy 
add a little commit... :innocent:
#[comment in UP-4735](https://issues.jasig.org/browse/UP-4735?focusedCommentId=39550&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-39550) VoiceOver is reacting differently on OS X and iOS and iOS differently between an iPhone and an iPad with le Selected Message close link. :grimacing: :sweat:
So I change from role="button" to role="link". Now every links and buttons can be activated with VoiceOver on OS X and IOS (iphone/ipad). :smile:

> A related issue on pager was found while testing the role="application" changes that may affect email preview.
https://issues.fluidproject.org/browse/FLUID-6081

It is not applicable in our case for the moment. in fluid latest, there is a title on every numbered links in the pagination of a grid table and it should spell something like "Page 2: Joe I Instructor - Fernando Horwitz" . 

In the case of emailPreview it can be both (my POV):

- a non sense order: spelling "page 2: message from C. Murphy to message from A. Cousquer"
- maybe a performance issue: js needs to parse each page messages to grab every page first and last messages item

for FLUID-6064, nice done 👍 @ChristianMurphy  
I couldn't personnaly imagine that the Fluid team will fix this so quickly. :rainbow:
Many many Thanks to @jobara. :ok_hand: :clap:

Well, now, we need to update to FLUID latest which won't be trivial. (update of resourceserver, remove rsf from preview.jsp, update from fluid 1.4 to Fluid 1.5 to Fluild latest), **we need really a plan**. 
Or perhaps the fluid Team may patch branches version1.4 and version1.5. i don't think that the issue fix FLUID-6064 can be backported easily on these branches, because the code has changed a lot.

ChristianMurphy commented on src/main/webapp/WEB-INF/jsp/preview.jsp:
```diff
@@ -66,7 +66,7 @@
                 <c:if test="${showConfigLink}">
                     <portlet:renderURL var="configUrl" portletMode="CONFIG"/>
-                    <a class="btn btn-primary pull-right" role="button" href="${ configUrl }"><i class="fa fa-gear" aria-hidden="true"></i> <spring:message code="preview.configure"/></a>
+                    <a class="btn btn-primary pull-right" role="link" href="${ configUrl }"><i class="fa fa-gear" aria-hidden="true"></i> <spring:message code="preview.configure"/></a>
```
I believe that the `<a>` tag has `role="link"` by default.

Reference: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_link_role

cousquer replied:
I definitly know that, but dunno if this came from CSS `transform: translate3D` double-inner off-canvas design, I was loosing the focus on a element in some old android, iPad and iPhone 4, so I have added both `role='link/button"` + tabindex on `<a>`

ChristianMurphy replied:
:thinking: interesting :thinking:
that happens with `role="button"` or role-less `<a>` but not with `role="link"`? and only on mobile devices?

ChristianMurphy commented:
> Well, now, we need to update to FLUID latest which won't be trivial. (update of resource server, remove rsf from `preview.jsp`, update from Fluid 1.4 to Fluid 1.5 to Fluid latest), **we need really a plan**.
Or perhaps the Fluid Team may patch branches version 1.4 and version 1.5. I don't think that the issue fix FLUID-6064 can be backported easily on these branches, because the code has changed a lot.

I am working on a backport of the patch to version 1.4 in https://github.com/fluid-project/infusion/pull/785
It would still be good to migrate to fluid version 2, @jobara mentioned he ***may*** have some time to poke around and offer some suggestions or help on a migration.

cousquer commented:
> I am working on a backport of the patch to version 1.4 in fluid-project/infusion#785
> It would still be good to migrate to fluid version 2, @jobara mentioned he may have some time to poke around and offer some suggestions or help on a migration.
Great! 👍 👍 👍

cousquer commented:
#https://github.com/Jasig/uPortal/commit/999aabaeaaac3157aa448bc8907f21a4097fb157

cousquer commented:
#
https://github.com/Jasig/uPortal/commit/999aabaeaaac3157aa448bc8907f21a4097fb157

cousquer commented:
##jasig/uPortal#681

cousquer commented:
commit 999aabaeaaac3157aa448bc8907f21a4097fb157 #681

cousquer commented:
Jasig/uPortal#681

cousquer commented:
@ChristianMurphy 
now with a screenreader (Voiceover or Talkback) or without it, it's easy to scroll up after reading a long message to close the selected message and go to email-list to open another message...

cousquer commented:
@ChristianMurphy 
see Special cases of Criterion 5.3 [A] (RGAA 3 2016)
> The ARIA API provides a mechanism to overload the native role of an HTML element to provide components. Thus, it is possible to use formatting tables to build lists:

```
<table role="list">
   <tr role="listitem">
        <td role="presentation"><img src="puce.png" alt=""/></td>
        <td role="presentation">lorem ipsum</td>
   </tr>
 …
</table>
```

> If this use is discouraged, it is nonetheless compliant. Since the array is not rendered as an array but as a list, it is not useful to report that it is a formatting table. 

[Criterion 5.3 [A] For each formatting table, is linearized content still comprehensible (excluding special cases)?](http://references.modernisation.gouv.fr/rgaa-accessibilite/criteres.html#tableaux)
[Special cases of the french transposition of WCAG2 (RGAA 3 2016) ](http://references.modernisation.gouv.fr/rgaa-accessibilite/cas-particuliers.html#cp-5-3)

And I want to keep CSS display:"table"... 😉

ChristianMurphy commented on src/main/webapp/WEB-INF/jsp/preview.jsp:
```diff
@@ -246,7 +246,7 @@
                     <table id="email-list-table" class="list-group" role="list">
                         <tr rsf:id="row:" class="email-row list-group-item" role="listitem">
-                           <td class="right-content-email-toggle">
+                           <td class="right-content-email-toggle" role="presentation">
```
Could the same look and behavior of table with aria roles

``` jsp
<table id="email-list-table" class="list-group" role="list">
    <tr rsf:id="row:" class="email-row list-group-item" role="listitem">
        <td class="right-content-email-toggle" role="presentation">
            ... (content here) ...
        </td>
    </tr>
</table>
```
 be achieved using list with an unordered list using `display: table-cell`?

``` jsp
<style>
    .table-cell {
        display: table-cell;
    }
</style>

<ul>
    <li rsf:id="row:" class="table-cell email-row">
        ... (content here) ...
    </li>
<ul>
```

The advantage of the change would be the DOM tree matching the accessibility tree, making maintenance easier.
The disadvantage of the change would be it uses CSS `display: table-cell;` which is considered bad practice.

cousquer replied:
@ChristianMurphy 
two things:
* We are in fluid in rsf way. so changing table tr to ul li is not the only things to do. if you do just that the email'list isn't rendered (I've tested that option).
* in fluid 1.1 (jQueryMobile view) there is [this](https://github.com/Jasig/email-preview/blob/506854d62aca7e1d43dc883869ebb8875ed2b515/src/main/webapp/js/email-browser.js#L592) in `email-browser.js`, And I wasn't sure of Fluid1.1's behavior. it seems to test and script `<table>` into a `<div>`. So with a `<table>` I'm confident in its behavior. and I need its CSS default display's behavior to let users zoom the email-list. On touch device it was working great whatever browser and OS. so -> keep `<table>` .
_(btw you can't zoom in email-list in google's Inbox native app, you can only zoom in the selected message...)_

ChristianMurphy replied:
Fair enough, if it is a pain to migrate, keep the design as is.

> btw you can't zoom in email-list in google's Inbox native app, you can only zoom in the selected message

Is that on IOS? On Android I can zoom in and out on all views.

cousquer replied:
> keep the design as is

yes

> Is that on IOS?

Yes on IOS in the native App Context (remember role=application thread...)
in IOS,  `IOS>Settings>accessibility>Set font Size to the maximum`
open App INBOX : email-list isn't displayed with big Size text
so try to two fingers zoom > **no zoom on email-list**

**IOS:**  0  _  **Google INBOX:**  -1

in IOS  `IOS>Settings>accessibility> enable Magnifier`
open App INBOX
in Email-list with the magnifier : a user can read about **30%** of email-list's text, he needs to **scroll on the left and on right** to read 100% of the box.
but when he scroll on the right or the left with the magnifier, **it fires the swipe events right || left**, so _the message is automatically archived || supressed and the message disapears..._

**IOSmagnifier:**  -1  _  **Google INBOX:** -1

> draw 😄

jobara commented:
@ChristianMurphy and @cousquer, I've tagged an [Infusion 1.4.1 release](https://github.com/fluid-project/infusion/releases/tag/infusion-1.4.1) which includes the changes from PR to adjust the use of `role="application"`.

ChristianMurphy commented:
Thanks @jobara!

cousquer commented:
Many many Thanks @jobara ! 

@ChristianMurphy, I'll try in the next weeks: 
1. PR update resourceserver : with everything uptodate
2. test with Infusion1.4.1
3. if OK (I'm really confident in point 3. 😄 )
4. let's do a visual track to :clap: &  🎉 @jobara  : 
  4.1 target point [3.2 Touch Target Size and Spacing.](https://www.w3.org/TR/mobile-accessibility-mapping/#touch-target-size-and-spacing)
   so too small square buttons :black_small_square: => rounded (9mm) buttons :black_circle: like [Keyline shapes of MD](https://material.google.com/style/icons.html#icons-product-icons) ?
  ( _it was something Marissa W. W. from Edinburgh University told me last june at Apereo NY_)
   @ChristianMurphy : :question: What do you think of that? ;question:

ChristianMurphy commented:
@cousquer Sounds good to me. :+1:

ChristianMurphy commented:
https://issues.jasig.org/browse/RESSERV-92 and https://issues.jasig.org/browse/RESSERV-93 have been opened to make Fluid 1.4.1 and 2.0.0 available in uPortal.

cousquer commented:
@bjagg @ChristianMurphy @jobara 
- Implement Material Design on main action buttons / links. (rounded btn)
- conformed to [3.2 Touch Target Size and Spacing of W3C mobile accessibility Draft ](https://www.w3.org/TR/mobile-accessibility-mapping/#touch-target-size-and-spacing).
- Solving this accessibility open issue in Bootstrap v4 with touch-sr-only class see [sr-only-*-up / sr-only-*-down utility classes for accessibility](https://github.com/twbs/bootstrap/issues/19954). the main problem came from VoiceOver rotor which is interacting with the graphical layer and highlights hidden text on a sr-hidden icon link. You need to know that Bootstrap sr-only class was created to display the skip to content/nav block at the beginning. So it was another use case than the hidden text inside a link
Test ok with VoiceOver and Talkback

ChristianMurphy commented:
Thanks @cousquer!
I went through and captured screenshots from each view in the portlet.

![screencapture-localhost-8080-uportal-p-email-preview-demo-ctf3-max-render-up-1486480352799](https://cloud.githubusercontent.com/assets/3107513/22697054/704991f4-ed0d-11e6-9dab-81e94fc6b03b.png)
**Overview**
It looks really good, I really like how the colors play together and the Material design look in general. :art: 

A couple low priority suggestions:

* It is unclear that the power button in the left panel is designed to close the right panel, since the right panel has a close button, this button may not be needed.
* Material design encourages the use of [bright solid colors](https://material.io/guidelines/style/color.html), the header could be updated to be a solid instead of a gradient.

---

![screencapture-localhost-8080-uportal-p-email-preview-demo-ctf3-max-render-up-1486480362272](https://cloud.githubusercontent.com/assets/3107513/22697056/705957e2-ed0d-11e6-8bf5-f86e9aadd4ed.png)
**Settings**

This view also looks good! :slightly_smiling_face: 

---

![screencapture-localhost-8080-uportal-p-email-preview-demo-ctf3-max-render-up-1486480385869](https://cloud.githubusercontent.com/assets/3107513/22697055/705946e4-ed0d-11e6-90ba-8f2e4d5e22f9.png)
**Email Preview**

I like the colorful icons on this view, paired with `title` used to create tooltip on hover. :+1: 

---

![screencapture-localhost-8080-uportal-p-email-preview-demo-ctf3-max-render-up-1486480405957](https://cloud.githubusercontent.com/assets/3107513/22697057/705b3b52-ed0d-11e6-9d50-af2e837935e5.png)
**Help Page**

This page is a bit bland, but help pages often are. :man_shrugging: 

---

![screencapture-localhost-8080-uportal-p-email-preview-demo-ctf3-max-render-up-1486480423581](https://cloud.githubusercontent.com/assets/3107513/22697058/705b8e5e-ed0d-11e6-842d-9499f6ff06f0.png)
**Multi-Selection View**

Nice use of a plain background, with a bright icon to make availible actions clear.

:warning: High Priority Suggestion: :warning: 

* There is no visual indication of what the checkbox next to "cancel" does, this could use a visible "select all" label.

Low Priority Suggestions:

* Apply Material checkbox style to select all.
* The multi-select menu seems to be a pixel or two too high, I can see the menu undernearth peeking out.
